robonomics_comm doesn't understand service messages from provider

The stack doesn't understand the messages described in here https://github.com/airalab/robonomics-tools/issues/1

```
[ERROR] [1566044183.469201]: Message validation error: required key not provided @ data['validatorFee']
[WARN] [1566044183.473185]: Message {'signature': '0x3820c99af1518a4c2bf13b79ab144c1451928e1ab9f361e4c506f29ea4e8f0e54bbe0b2f7f401c269f43f14ed009329a13ecbf9a2c9e0d0806bffabf654fc5c81b', 'order': '0x6c9cb72e1a7cd5ec94f9dc2fac64082278d8788a190c13e6d6dca138250082ff', 'accepted': 0} is not valid Ask, Bid or Result ipfs message
```

--- a/aira_graph/src/aira_graph/nodes.py
+++ b/aira_graph/src/aira_graph/nodes.py
@@ -1,5 +1,5 @@
-from . import aira_graph
+from .aira_graph import AIRAGraph
 
 
 def aira_graph_node():
-    aira_graph.AIRAGraph().spin()
+    AIRAGraph().spin()

--- a/ethereum_common/src/ethereum_common/nodes.py
+++ b/ethereum_common/src/ethereum_common/nodes.py
@@ -3,18 +3,18 @@
 # ethereum common related nodes.
 #
 
-from . import erc20_services
-from . import eth_services
-from . import signer
+from .erc20_services import ERC20Services
+from .eth_services import ETHServices
+from .signer import Signer
 
 
 def erc20_node():
-    erc20_services.ERC20Services().spin()
+    ERC20Services().spin()
 
 
 def eth_node():
-    eth_services.ETHServices().spin()
+    ETHServices().spin()
 
 
 def signer_node():
-    signer.Signer().spin()
+    Signer().spin()

--- a/ipfs_common/src/ipfs_common/ipfs_channel.py
+++ b/ipfs_common/src/ipfs_common/ipfs_channel.py
@@ -3,9 +3,9 @@
 # Robonomics information channels support node.
 #
 
-from robonomics_msgs.msg import Demand, Offer, Result
+from robonomics_msgs.msg import Demand, Offer, Result, AddedOrderFeedback, AddedPendingTransactionFeedback
 from robonomics_msgs.messageValidator import convertMessage
-from robonomics_msgs.robonomicsMessageUtils import offer2dict, demand2dict, res2dict
+from robonomics_msgs.robonomicsMessageUtils import offer2dict, demand2dict, res2dict, addedOrderFeedback2dict, addedPendingTransactionFeedback2dict
 from .pubsub import publish, subscribe
 from urllib.parse import urlparse
 from threading import Thread
@@ -26,10 +26,19 @@ class IPFSChannel:
         self.incoming_offer  = rospy.Publisher('incoming/offer',  Offer, queue_size=10)
         self.incoming_demand = rospy.Publisher('incoming/demand', Demand, queue_size=10)
         self.incoming_result = rospy.Publisher('incoming/result', Result, queue_size=10)
+        self.incoming_added_order_feedback = rospy.Publisher('incoming/feedback/added_order', AddedOrderFeedback, queue_size=10)
+        self.incoming_added_pending_transaction_feedback = rospy.Publisher('incoming/feedback/added_pending_transaction', AddedPendingTransactionFeedback, queue_size=10)
 
-        rospy.Subscriber('eth/sending/offer',  Offer,  lambda m: publish(self.ipfs_client, self.lighthouse, offer2dict(m)))
-        rospy.Subscriber('eth/sending/demand', Demand, lambda m: publish(self.ipfs_client, self.lighthouse, demand2dict(m)))
-        rospy.Subscriber('eth/sending/result', Result, lambda m: publish(self.ipfs_client, self.lighthouse, res2dict(m)))
+        rospy.Subscriber('eth/sending/offer',  Offer,
+                         lambda m: publish(self.ipfs_client, self.lighthouse, offer2dict(m)))
+        rospy.Subscriber('eth/sending/demand', Demand,
+                         lambda m: publish(self.ipfs_client, self.lighthouse, demand2dict(m)))
+        rospy.Subscriber('eth/sending/result', Result,
+                         lambda m: publish(self.ipfs_client, self.lighthouse, res2dict(m)))
+        rospy.Subscriber('eth/sending/feedback/added_order', AddedOrderFeedback,
+                         lambda m: publish(self.ipfs_client, self.lighthouse, addedOrderFeedback2dict(m)))
+        rospy.Subscriber('eth/sending/feedback/added_pending_transaction', AddedPendingTransactionFeedback,
+                         lambda m: publish(self.ipfs_client, self.lighthouse, addedPendingTransactionFeedback2dict(m)))
 
     def spin(self):
         '''
@@ -48,6 +57,12 @@ class IPFSChannel:
                     elif isinstance(converted, Result):
                         # rospy.logwarn('DEBUG: Publish valid Result message %s', converted)
                         self.incoming_result.publish(converted)
+                    elif isinstance(converted, AddedOrderFeedback):
+                        # rospy.logwarn('DEBUG: Publish valid AddedOrderFeedback message %s', converted)
+                        self.incoming_added_order_feedback.publish(converted)
+                    elif isinstance(converted, AddedPendingTransactionFeedback):
+                        # rospy.logwarn('DEBUG: Publish valid AddedPendingTransactionFeedback message %s', converted)
+                        self.incoming_added_pending_transaction_feedback.publish(converted)
 
         Thread(target=channel_thread, daemon=True).start()
         rospy.spin()

--- a/ipfs_common/src/ipfs_common/nodes.py
+++ b/ipfs_common/src/ipfs_common/nodes.py
@@ -1,10 +1,10 @@
-from . import ipfs_channel
-from . import ipfs_common
+from .ipfs_channel import IPFSChannel
+from .ipfs_common import IPFSCommon
 
 
 def ipfs_channel_node():
-    ipfs_channel.IPFSChannel().spin()
+    IPFSChannel().spin()
 
 
 def ipfs_common_node():
-    ipfs_common.IPFSCommon().spin()
+    IPFSCommon().spin()

--- a/robonomics_control/src/robonomics_control/nodes.py
+++ b/robonomics_control/src/robonomics_control/nodes.py
@@ -3,7 +3,7 @@
 # Robonomics control related nodes.
 #
 
-from . import distribution
+from .distribution import Distribution
 
 def distribution_node():
-    distribution.Distribution().spin()
+    Distribution().spin()

--- a/robonomics_liability/src/robonomics_liability/nodes.py
+++ b/robonomics_liability/src/robonomics_liability/nodes.py
@@ -3,23 +3,23 @@
 # Robonomics control related nodes.
 #
 
-from . import listener
-from . import executor
-from . import LiabilityExecutionsPersistence
-from . import player_service
+from .listener import Listener
+from .executor import Executor
+from .LiabilityExecutionsPersistence import LiabilityExecutionsPersistence
+from .player_service import PlayerService
 
 
 def listener_node():
-    listener.Listener().spin()
+    Listener().spin()
 
 
 def executor_node():
-    executor.Executor().spin()
+    Executor().spin()
 
 
 def persistence_node():
-    LiabilityExecutionsPersistence.LiabilityExecutionsPersistence().spin()
+    LiabilityExecutionsPersistence().spin()
 
 
 def player_service_node():
-    player_service.PlayerService().spin()
+    PlayerService().spin()

--- a/robonomics_liability/tests/testMessages.py
+++ b/robonomics_liability/tests/testMessages.py
@@ -1,4 +1,4 @@
-from robonomics_msgs.msg import Demand, Offer, Result
+from robonomics_msgs.msg import Demand, Offer, Result, AddedOrderFeedback, AddedPendingTransactionFeedback
 from ethereum_common.msg import Address, UInt256
 from ipfs_common.msg import Multihash
 from binascii import unhexlify
@@ -101,6 +101,37 @@ def getValidRes():
     r.success = validResDict['success']
     r.signature = unhexlify(validResDict['signature'].encode('utf-8'))
     return r
+
+
+validAddedOrderFeedbackDict = {
+    'signature': '3820c99af1518a4c2bf13b79ab144c1451928e1ab9f361e4c506f29ea4e8f0e54bbe0b2f7f401c269f43f14ed009329a13ecbf9a2c9e0d0806bffabf654fc5c81b',
+    'order': '6c9cb72e1a7cd5ec94f9dc2fac64082278d8788a190c13e6d6dca138250082ff',
+    'accepted': 0
+}
+
+
+def getValidAddedOrderFeedback():
+    f = AddedOrderFeedback()
+
+    f.order = unhexlify(validAddedOrderFeedbackDict['order'].encode('utf-8'))
+
+    f.accepted = UInt256()
+    f.accepted.uint256 = str(validAddedOrderFeedbackDict['accepted'])
+
+    f.signature = unhexlify(validAddedOrderFeedbackDict['signature'].encode('utf-8'))
+    return f
+
+
+validAddedPendingTransactionFeedbackDict = {
+    'tx': '6c9cb72e1a7cd5ec94f9dc2fac64082278d8788a190c13e6d6dca138250082ff',
+}
+
+
+def getValidAddedPendingTransactionFeedback():
+    f = AddedPendingTransactionFeedback()
+
+    f.tx = unhexlify(validAddedPendingTransactionFeedbackDict['tx'].encode('utf-8'))
+    return f
 
 
 #invalid data

--- a/robonomics_liability/tests/test_infochan.py
+++ b/robonomics_liability/tests/test_infochan.py
@@ -1,8 +1,7 @@
 #!/usr/bin/env python
 
 import rostest, sys, unittest, rospy, time
-from ipfs_common import ipfs_channel
-from robonomics_msgs.msg import Demand, Offer, Result
+from robonomics_msgs.msg import Demand, Offer, Result, AddedOrderFeedback, AddedPendingTransactionFeedback
 from robonomics_msgs.robonomicsMessageUtils import offer2dict, demand2dict, res2dict
 import testMessages
 
@@ -18,13 +17,27 @@ class TestInfochan(unittest.TestCase):
         self._infochan_published_ask_success = False
         self._infochan_published_bid_success = False
         self._infochan_published_res_success = False
+        self._infochan_published_added_order_feedback_success = False
+        self._infochan_published_added_pending_transaction_feedback_success = False
 
-        rospy.Subscriber('/liability/infochan/incoming/demand', Demand, self.infochan_published_ask_handler)
-        rospy.Subscriber('/liability/infochan/incoming/offer',  Offer,  self.infochan_published_bid_handler)
-        rospy.Subscriber('/liability/infochan/incoming/result', Result, self.infochan_published_result_handler)
+        rospy.Subscriber('/liability/infochan/incoming/demand',
+                         Demand, self.infochan_published_ask_handler)
+        rospy.Subscriber('/liability/infochan/incoming/offer',
+                         Offer,  self.infochan_published_bid_handler)
+        rospy.Subscriber('/liability/infochan/incoming/result',
+                         Result, self.infochan_published_result_handler)
+        rospy.Subscriber('/liability/infochan/incoming/feedback/added_order',
+                         AddedOrderFeedback, self.infochan_published_added_order_feedback_handler)
+        rospy.Subscriber('/liability/infochan/incoming/feedback/added_pending_transaction',
+                         AddedPendingTransactionFeedback, self.infochan_published_added_pending_transaction_feedback_handler)
+
         self.infochan_Ask_subscriber_topic = rospy.Publisher('/liability/infochan/eth/sending/demand', Demand, queue_size=10)
         self.infochan_Bid_subscriber_topic = rospy.Publisher('/liability/infochan/eth/sending/offer',  Offer,  queue_size=10)
         self.infochan_Res_subscriber_topic = rospy.Publisher('/liability/infochan/eth/sending/result', Result, queue_size=10)
+        self.infochan_Added_Order_Feedback_subscriber_topic = rospy.Publisher('/liability/infochan/eth/sending/feedback/added_order',
+                                                                              AddedOrderFeedback, queue_size=10)
+        self.infochan_Added_Pending_Transaction_Feedback_subscriber_topic = rospy.Publisher('/liability/infochan/eth/sending/feedback/added_pending_transaction',
+                                                                              AddedPendingTransactionFeedback, queue_size=10)
 
 
     def test_demand2dict(self):
@@ -71,6 +84,30 @@ class TestInfochan(unittest.TestCase):
         while not rospy.is_shutdown() and not self._infochan_published_res_success and time.time() < timeout_t:
             time.sleep(0.1)
         self.assert_(self._infochan_published_res_success)
+
+    def infochan_published_added_order_feedback_handler(self, feedback):
+        self.assertEqual(testMessages.getValidAddedOrderFeedback(), feedback)
+        self._infochan_published_added_order_feedback_success = True
+
+    def test_added_order_feedback_is_published_on_topic(self):
+        time.sleep(3) #infochan node subscribers may be not registered in master
+        self.infochan_Added_Order_Feedback_subscriber_topic.publish(testMessages.getValidAddedOrderFeedback())
+        timeout_t = time.time() + 60.0
+        while not rospy.is_shutdown() and not self._infochan_published_added_order_feedback_success and time.time() < timeout_t:
+            time.sleep(0.1)
+        self.assert_(self._infochan_published_added_order_feedback_success)
+
+    def infochan_published_added_pending_transaction_feedback_handler(self, feedback):
+        self.assertEqual(testMessages.getValidAddedPendingTransactionFeedback(), feedback)
+        self._infochan_published_added_pending_transaction_feedback_success = True
+
+    def test_added_pending_transaction_feedback_is_published_on_topic(self):
+        time.sleep(3) #infochan node subscribers may be not registered in master
+        self.infochan_Added_Pending_Transaction_Feedback_subscriber_topic.publish(testMessages.getValidAddedPendingTransactionFeedback())
+        timeout_t = time.time() + 60.0
+        while not rospy.is_shutdown() and not self._infochan_published_added_pending_transaction_feedback_success and time.time() < timeout_t:
+            time.sleep(0.1)
+        self.assert_(self._infochan_published_added_pending_transaction_feedback_success)
 
 
 if __name__ == '__main__':

--- a/robonomics_liability/tests/test_messageValidator.py
+++ b/robonomics_liability/tests/test_messageValidator.py
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
 
 import unittest, rostest, sys
-from robonomics_msgs.msg import Result, Offer, Demand
+from robonomics_msgs.msg import Result, Offer, Demand, AddedOrderFeedback, AddedPendingTransactionFeedback
 from robonomics_msgs import messageValidator
 import testMessages
 
@@ -57,25 +57,13 @@ class TestMessageValidator(unittest.TestCase):
         f = messageValidator.convertMessage(testMessages.invalidRes2Dict)
         assert f is None
 
-    def test_InvalidAsk3(self):
-        f = messageValidator.convertMessage(testMessages.invalidAsk3Dict)
-        assert f is None
+    def test_ValidAddedOrderFeedback(self):
+        f = messageValidator.convertMessage(testMessages.validAddedOrderFeedbackDict)
+        assert isinstance(f, AddedOrderFeedback)
 
-    def test_InvalidBid3(self):
-        f = messageValidator.convertMessage(testMessages.invalidBid3Dict)
-        assert f is None
-
-    def test_InvalidRes3(self):
-        f = messageValidator.convertMessage(testMessages.invalidRes3Dict)
-        assert f is None
-
-    def test_InvalidAsk4(self):
-        f = messageValidator.convertMessage(testMessages.invalidAsk4Dict)
-        assert f is None
-
-    def test_InvalidBid4(self):
-        f = messageValidator.convertMessage(testMessages.invalidBid4Dict)
-        assert f is None
+    def test_ValidAddedPendingTransactionFeedback(self):
+        f = messageValidator.convertMessage(testMessages.validAddedPendingTransactionFeedbackDict)
+        assert isinstance(f, AddedPendingTransactionFeedback)
 
 
 if __name__ == '__main__':

--- a/robonomics_msgs/CMakeLists.txt
+++ b/robonomics_msgs/CMakeLists.txt
@@ -54,6 +54,8 @@ add_message_files(
         Demand.msg
         Offer.msg
         Result.msg
+        AddedOrderFeedback.msg
+        AddedPendingTransactionFeedback.msg
 )
 
 ## Generate services in the 'srv' folder

--- a/robonomics_msgs/msg/AddedOrderFeedback.msg
+++ b/robonomics_msgs/msg/AddedOrderFeedback.msg
@@ -1,0 +1,6 @@
+# Order hash
+uint8[32]               order       # order hash
+# Accepted
+ethereum_common/UInt256 accepted    # int
+# signature
+uint8[]                 signature   # hex

--- a/robonomics_msgs/msg/AddedPendingTransactionFeedback.msg
+++ b/robonomics_msgs/msg/AddedPendingTransactionFeedback.msg
@@ -1,0 +1,2 @@
+# Transaction hash
+uint8[32]               tx          # tx hash

--- a/robonomics_msgs/src/robonomics_msgs/messageValidator.py
+++ b/robonomics_msgs/src/robonomics_msgs/messageValidator.py
@@ -3,7 +3,9 @@
 # Robonomics Demand/Offer/Result ipfs message converter
 #
 
-from .robonomicsMessageUtils import offer2dict, demand2dict, res2dict, dict2offer, dict2demand, dict2res
+from .robonomicsMessageUtils import offer2dict, demand2dict, res2dict, \
+    dict2offer, dict2demand, dict2res, \
+    dict2addedOrderFeedback, dict2addedPendingTransactionFeedback
 from robonomics_msgs.msg import Demand, Offer, Result
 import voluptuous as v
 import rospy
@@ -11,16 +13,28 @@ import rospy
 
 @v.message('wrong hexadecimal field value')
 def isHexIntNotZeroCanBeEmpty(arg):
-    if arg and not int(arg, 16) > 0:
-        raise v.Invalid('value is zero')
+    if arg:
+        try:
+            int_arg = int(arg, 16)
+            if not int_arg > 0:
+                raise v.Invalid('value is zero')
+            else:
+                return hex(int_arg)[2:]
+        except ValueError as e:
+            raise v.Invalid(e)
     return arg
 
 
-@v.message('wrong decimal field value')
-def isPositiveDecIntCanBeEmpty(arg):
-    if arg and not arg >= 0:
-        raise v.Invalid('value is negative')
-    return arg
+@v.message('wrong TxHash field value')
+def isTxHash(arg):
+    try:
+        int_arg = int(arg, 16)
+        if not int_arg > 0:
+            raise v.Invalid('value is zero')
+        else:
+            return hex(int_arg)[2:]
+    except ValueError as e:
+        raise v.Invalid(e)
 
 
 isIpfsBase58Hash = v.All(str, v.Length(min=46, max=46), v.Match(r'^[a-zA-Z0-9]+$'))
@@ -51,7 +65,7 @@ schemaAskBid = v.All(
         v.Required('cost'): v.All(int),
         v.Required('deadline'): v.All(int),
         v.Required('sender'): isAddress,
-        v.Required('nonce'): isPositiveDecIntCanBeEmpty(),
+        v.Required('nonce'): v.All(int),
         v.Required('signature'): isHexIntNotZeroCanBeEmpty()
     })
 )
@@ -63,9 +77,21 @@ schemaResult = v.Schema({
     v.Required('signature'): isHexIntNotZeroCanBeEmpty()
 })
 
-schemaAskBidResult = v.Any(
+schemaAddedOrderFeedback = v.Schema({
+    v.Required('signature'): isHexIntNotZeroCanBeEmpty(),
+    v.Required('order'): isTxHash(),
+    v.Required('accepted'): v.All(int)
+})
+
+schemaAddedPendingTransactionFeedback = v.Schema({
+    v.Required('tx'): isTxHash(),
+})
+
+schemaAskBidResultReport = v.Any(
     schemaAskBid,
-    schemaResult
+    schemaResult,
+    schemaAddedOrderFeedback,
+    schemaAddedPendingTransactionFeedback
 )
 
 
@@ -101,7 +127,7 @@ def isResultFieldsCorrect(msg):
 
 def msg_is_offer_demand_result(abr_msg):
     try:
-        return schemaAskBidResult(abr_msg)
+        return schemaAskBidResultReport(abr_msg)
     except v.MultipleInvalid as e:
         rospy.logerr("Message validation error: %s", str(e))
         return None
@@ -116,9 +142,15 @@ def convertMessage(ipfsMessage):
         elif 'lighthouseFee' in validatedBySchema:
             # rospy.logwarn('DEBUG: Message %s is valid Bid ipfs message', ipfsMessage)
             return dict2offer(validatedBySchema)
-        else:
+        elif 'accepted' in validatedBySchema:
+            # rospy.logwarn('DEBUG: Message %s is valid AddedOrderFeedback ipfs message', validatedBySchema)
+            return dict2addedOrderFeedback(validatedBySchema)
+        elif 'tx' in validatedBySchema:
+            # rospy.logwarn('DEBUG: Message %s is valid AddedPendingTransactionFeedback ipfs message', validatedBySchema)
+            return dict2addedPendingTransactionFeedback(validatedBySchema)
+        elif 'liability' in validatedBySchema:
             # rospy.logwarn('DEBUG: Message %s is valid Result ipfs message', ipfsMessage)
             return dict2res(validatedBySchema)
 
-    rospy.logwarn('Message %s is not valid Ask, Bid or Result ipfs message', ipfsMessage)
+    rospy.logwarn('Message %s is not valid Ask, Bid, Result or Feedback ipfs message', ipfsMessage)
     return None

--- a/robonomics_msgs/src/robonomics_msgs/robonomicsMessageUtils.py
+++ b/robonomics_msgs/src/robonomics_msgs/robonomicsMessageUtils.py
@@ -4,7 +4,7 @@
 #
 from web3 import Web3
 from web3.auto import w3
-from robonomics_msgs.msg import Demand, Offer, Result
+from robonomics_msgs.msg import Demand, Offer, Result, AddedOrderFeedback, AddedPendingTransactionFeedback
 from eth_account.messages import defunct_hash_message
 import multihash
 from web3.utils.normalizers import (
@@ -139,6 +139,20 @@ def res2dict(r):
     }
 
 
+def addedOrderFeedback2dict(f):
+    return {
+        'order': hexlify(f.order).decode('utf-8'),
+        'accepted': int(f.accepted.uint256),
+        'signature': hexlify(f.signature).decode('utf-8')
+    }
+
+
+def addedPendingTransactionFeedback2dict(f):
+    return {
+        'tx': hexlify(f.tx).decode('utf-8')
+    }
+
+
 def dict2demand(m):
     msg = Demand()
 
@@ -172,7 +186,7 @@ def dict2demand(m):
     msg.nonce = UInt256()
     msg.nonce.uint256 = str(m['nonce'])
 
-    msg.signature = unhexlify(m['signature'].encode('utf-8'))
+    msg.signature = unhexlify(m['signature'].encode('utf-8')) if m['signature'] else []
     return msg
 
 
@@ -209,7 +223,7 @@ def dict2offer(m):
     msg.nonce = UInt256()
     msg.nonce.uint256 = str(m['nonce'])
 
-    msg.signature = unhexlify(m['signature'].encode('utf-8'))
+    msg.signature = unhexlify(m['signature'].encode('utf-8')) if m['signature'] else []
     return msg
 
 
@@ -223,5 +237,24 @@ def dict2res(m):
     msg.result.multihash = m['result']
 
     msg.success = m['success']
+    msg.signature = unhexlify(m['signature'].encode('utf-8')) if m['signature'] else []
+    return msg
+
+
+def dict2addedOrderFeedback(m):
+    msg = AddedOrderFeedback()
+
+    msg.order = unhexlify(m['order'].encode('utf-8'))
+
+    msg.accepted = UInt256()
+    msg.accepted.uint256 = str(m['accepted'])
+
     msg.signature = unhexlify(m['signature'].encode('utf-8'))
+    return msg
+
+
+def dict2addedPendingTransactionFeedback(m):
+    msg = AddedPendingTransactionFeedback()
+
+    msg.tx = unhexlify(m['tx'].encode('utf-8'))
     return msg

